File: lc/coreprep/implementation/practice_iterations/fast_and_slow_pointers_5/HappyNumber.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(logn) / O(1)
public class HappyNumber {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        System.out.println(isHappy(n));
      }
    }
  }

  private static boolean isHappy(int n) {
    Function<Integer, Integer> digitSquareSum = (Integer m) -> {
      int sum = 0;

      while (m > 0) {
        sum += (m % 10) * (m % 10);
        m /= 10;
      }
      return sum;
    };

    int fast = n, slow = n;
    do {
      fast = digitSquareSum.apply(digitSquareSum.apply(fast));
      slow = digitSquareSum.apply(slow);
    } while (fast != slow);

    return slow == 1;
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/fast_and_slow_pointers_2/happy_number.cpp
#include <algorithm>
#include <iostream>

using namespace std;

bool is_happy(int n) {
  auto digit_square_sum = [](int m) {
    int sum = 0;

    while (m) {
      sum += (m % 10) * (m % 10);
      m /= 10;
    }

    return sum;
  };

  int fast = n, slow = n;

  do {
    fast = digit_square_sum(digit_square_sum(fast));
    slow = digit_square_sum(slow);
  } while (fast != slow);

  return slow == 1;
}

// O(logn) / O(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;
    cout << (is_happy(n) ? "true" : "false") << "\n";
  }

  return 0;
}